### PR TITLE
syncer: Add Close and ensure goroutines exit

### PR DESCRIPTION
**Pull request by lukechampine**
See https://github.com/SiaFoundation/renterd/pull/1407

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -276,6 +276,8 @@ func (s *Syncer) runPeer(p *Peer) error {
 	}()
 
 	inflight := make(chan struct{}, s.config.MaxInflightRPCs)
+	var wg sync.WaitGroup
+	defer wg.Done()
```
Comment (peterjan) [inline syncer/syncer.go:280]: This should be `defer wg.Wait()`

**Review by peterjan — COMMENTED** [review]
I didn't immediately see how this PR was ensuring the `syncLoop` exits, would it clean things up if we added a `wg` to the `syncer` and then use that in all loops and replace the `<-errChan` with `wg.Wait` right in `Run`?

File: syncer/syncer.go (hunk 1 above)
Comment (lukechampine) [reply syncer/syncer.go:280]: whoops, you're right

**Comment by lukechampine** [comment]
I think that would make the error handling more annoying; instead of sending the errors down a channel, you would need separate variables or a slice or something.

**Review by peterjan — DISMISSED** [review]

**Comment by peterjan** [comment]
> I think that would make the error handling more annoying; instead of sending the errors down a channel, you would need separate variables or a slice or something.

I'd keep the err channel but remove the two consecutive blocks where we don't care about the `err` being returned anyway. Anyway, doesn't matter, LGTM 👍

File: syncer/syncer.go [hunk 2]
```diff
@@ -632,7 +628,7 @@ func (s *Syncer) syncLoop() error {
 // Run spawns goroutines for accepting inbound connections, forming outbound
 // connections, and syncing the blockchain from active peers. It blocks until an
 // error occurs, upon which all connections are closed and goroutines are
-// terminated. To gracefully shutdown a Syncer, close its net.Listener.
+// terminated.
 func (s *Syncer) Run() error {
```
Comment (n8maninger) [inline syncer/syncer.go:632]: I think you can get rid of the `shutdownCtx` and `shutdownCtxCancel` fields by moving it here. Has, essentially, the same effect now that the waitgroup is there and is more idiomatic.

```go
func (s *Syncer) Run(ctx context.Context) error {
	errChan := make(chan error)
	go func() { errChan <- s.acceptLoop(ctx) }()
	go func() { errChan <- s.peerLoop(ctx) }()
	go func() { errChan <- s.syncLoop(ctx) }()
	err := <-errChan
        // ....
}
```

**Review by n8maninger — DISMISSED** [review]

**Review by n8maninger — APPROVED** [review]

**Pull request merged**